---
title: "Graphing Cytoband Data with RCytoGPS"
author: "Dwayne Tally"
date: "5/1/2020"
output:
  html_document: default
  pdf_document: default
  word_document: default
---
DOWNLOAD UMPIRE FROM R FORGE

```{r opts, echo=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=5)
```
# Introduction
Modern biological experiments are increasingly producing interesting binary matrices. These may represent the presence or absence of specific gene mutations, copy number variants, microRNAs, or other molecular or clinical phenomena. We have recently developed a tool, CytoGPS [^Abrams and colleagues], that converts conventional karyotypes from the standard text-based notation (the International Standard for Human Cytogenetic Nomenclature; ISCN) into a binary vector with three bits (loss, gain, or fusion) per cytoband, which we call the “LGF model”. To complement the website tool we have developed RCytoGPS as a package to help extract format and analyze genetic data processed by CytoGPS.org. The parsing JSON script is able to parse any JSON file extracted from CytoGPS.org and format them into an s4 object to help organizes the information. Also, a variety of idiogram visualizations are available to help users better interpret their cytogenetic data using frequency.

## Setup
First, you need to load the necassary library to extract data from JSON files.
```{r Setup}
library(RCytoGPS)
```
# Extracting JSON data and formatting to LGF model
readLGF is a function built within the RCytoGPS package. The function readLGF takes the parameters of a specific file name, from a specified directory/folder, and extracts and formats the JSON dataset into a nested list. This is done by using several built in functions called extractOneLGF to parse the data obtain from the JSON(s) file, and to format the data into a binary vector. Idioformat which takes the LGF output of extractOneLGF and produces the frequency of the cytogenetic events on the chromosome. Another function is rnames to build unique IDs based on the karyotype elements extracted. readLGF also impliments a class to construct the s4 organization method. Alternatively readLGF will also accept all JSON files from a specified directory if the file name is not specified.
```{r Extracting JSON data and formating to LGF model}
wd <-  system.file("Examples/JSONfiles", package = "RCytoGPS")
temp <- readLGF(folder = wd)
```
# Extracting the cytoband locations, and the frequency data
The purpose of this code is to extract and concatinate the cytoband location and the data frequency of the LGF model into a dataframe that the function CytobandData could be applied to to generate an s4 object. The new S4 object could then be used to generate multiple images.
```{r Extracting and formating Idiograph data}
cytoData <- data.frame(temp[["CL"]],temp[["frequency"]])
```

# Turning CytoData into an S4 Object
The purpose of this code is to transfrom the CytoData, into an S4 object using the function cytobandData located in the RcytoGPS package. CytobandData is used to obtain more detailed information of the CytoData such as looking for specific arms or bands of a cytogenetic abnormality. The newly aquired s4 object will then be used in generating Idiographs. 
```{r Turning CytoData into an S4 Object}
bandData <- CytobandData(cytoData)

```
## Generating Graphs

# Plotting Cytoband Data Along the Genome
The first visual is to show the user that they can generate graphs using different data sets obtained from different karyotypes. This is important because it allows an easier visual comparison between the same cytogenetic abnormality but using different karyotype data. The arguments for the function is the dataset, a specified column name or column index from bandData, and the col or color you want the graphs to be.
```{r gb, fig.cap="Cytoband level data along the genome.", fig.width=10, fig.height=5}
opar <- par(mfrow=c(2,1))
barplot(bandData, what = 8, col = "forestgreen")
par(opar)
```

# Single Chromosome Plots
Note: “Horizontal” and “vertical” refer to the direction of the chromosome with its cytoband shading. 

## Plotting Cytoband-Level Data Along One Chromosome
This graph allows you to simultaneously compare multiple cytogenetic events. The parameters are similar to the previous graph, except you can have a variable that stores the column names for the columns you want to analyze. You can also make another variable to hold the cytogenetic event from multiple karyotype datasets, but you have to specify the columns through the variable index and you have to specify chromosome you want to look at.
```{r v.plot2Chrom, fig.width=8, fig.height=5,fig.cap = "Vertical stacked barplot of LGF frequencies on chromosome 3 for type A samples."}
Acolumns <- names(temp[["frequency"]])
multiEvents <- list(Loss = Acolumns[grep("Loss", Acolumns)], Gain = Acolumns[grep("Gain", Acolumns)], Fusion = Acolumns[grep("Fusion", Acolumns)])
image(bandData, what = Acolumns[1:3], chr = 14)
```
Similar graph as the previous that takes in similar parameters, with difference is that the graph is presented along the x axis instead of the y axis. The purpose is for user preference, if they want to see a visual graph on the x axis instead of the y axis.
```{r h.plot2Chrom, fig.width=5, fig.height=8, fig.cap = "Horizontal stacked barplot of LGF frequencies on chromosome 3 for type A samples."}
image(bandData, Acolumns[1:3], chr = 14, horiz = TRUE)
```

# Idiograms

## One Data Column
This is the first of the Idiogram graphs you can make. The purpose of this graph is to have a visualization of the chromosomes as well as a barplot of the cytogenetic abnormalities attached to observe and possibly identify patterns. This function takes similar arguemnts to the previous except to change the color you have to use the argument pal.
```{r makeid, fig.width=10, fig.height=8, fig.cap="Idiogram for one data columm."}
image(bandData, what = Acolumns[1], chr = "all", pal = "orange")
```

## Contrasting Two Data Columns
This graph shows the user to compare and contrast two cytogenetic events simultaneously. The user can compare cytogenetic events from the same karyotypes or different ones. The arguments are similar to the previous ones, except you provide a list containing exactly two columns with the column names of the cytogenetic abnormalities you want to observe, and pal has to take in two different colors.
```{r biid, fig.width=10, fig.height=8, fig.cap="Idiogram to contrast two data columms."}
image(bandData, what = Acolumns[2:3], chr = "all", 
      pal=c("orange", "forestgreen"))
```

# Gallery
To see all possible visuals please go to our gallery for images.

# Appendix
SessionInfo() is used to get information on the local machine that ran the code. This is mostly for reproducibility purposes.
```{r}
sessionInfo()

```